### PR TITLE
[RFR] Add error page, update test, better management of invitation/creation channel

### PR DESCRIPTION
**Pull request by lcordier42**

### CODE REVIEW

File: pages/index.tsx [hunk 1]
```diff
@@ -18,6 +18,9 @@ const IndexPage: React.SFC<{
     if (!token) {
         return <Error statusCode={404} />;
     }
+    if (role === "candidate" && candidateName !== undefined) {
+        candidateName = undefined;
+    }
```
Comment (lcordier42) [inline pages/index.tsx]: Before that, when a candidate was logged and got candidate_id in query, it tried to create a channel automatically, is it better to get an error, or just remove candidateName (this var is a string who contain the name of the candidate we want to chat) like i do it here?

File: __tests__/sample.test.ts [hunk 2]
```diff
@@ -3,14 +3,14 @@ import * as puppeteer from "puppeteer";
 
 dotenv.config();
 
-let browserCandidate: puppeteer.Browser;
-let pageCandidate: puppeteer.Page;
+let candidateBrowser: puppeteer.Browser;
+let candidatePage: puppeteer.Page;
```
Comment (lcordier42) [inline __tests__/sample.test.ts:8]: I harmonized the var name with chat.test.ts

File: components/ChatApp.tsx [hunk 3]
```diff
@@ -279,29 +279,38 @@ export class ChatApp extends React.Component<
...
         if (this.props.candidateName !== undefined) {
+            const channelName =
+                this.props.username + " - " + this.props.candidateName;
             // Si un candidate a été invité
             const previousChannel = this.channel || undefined;
-            try {
+            // false === channel non existant, true === channel déjà crée
+            let created = false;
+            // @ts-ignore
+            const paginator = await this.chatClient.getSubscribedChannels();
```
Comment (lcordier42) [inline components/ChatApp.tsx]: Here, ts gives me an error, it thinks that getSubscribedChannels need an argument, but in twilio doc, it got no args, so i silent this warning

File: __tests__/chat.test.ts [hunk 4]
```diff
@@ -160,6 +158,48 @@ test("An admin can continue an existing chat", async () => {
...
 
+test("An employer can invite a candidate on chat", async () => {
+    // connecting employer to candidates page
+    await employerPage.goto(
+        `http://localhost:${
+            process.env.PORT
+        }/candidate?user_id=10&role=employer`,
+    );
+    // invite a candidate
+    await employerPage.click('[name="claire"]');
```
Comment (brikou) [inline __tests__/chat.test.ts]: is it an input, a button or ?, use explicit selector

File: components/CandidatesNav.tsx [hunk 5]
```diff
@@ -7,7 +7,7 @@ export const CandidatesNav = (props: any) => (
     <div>
         {candidates.map((candidat: string, id) => (
             <li key={id}>
-                <button>
+                <button name={candidat}>
```
Comment (brikou) [inline components/CandidatesNav.tsx]: always add type="button", otherwise it is by default type="submit"

File: components/ChatApp.tsx (hunk 3 above)
Comment (brikou) [reply components/ChatApp.tsx]: try 
```js
await this.chatClient.getSubscribedChannels(undefined);
```

File: components/ChatApp.tsx [hunk 6]
```diff
@@ -279,29 +279,36 @@ export class ChatApp extends React.Component<
...
+            let i;
+            for (i = 0; i < paginator.items.length; i++) {
```
Comment (brikou) [inline components/ChatApp.tsx]: use Array.some

File: components/ChatApp.tsx [hunk 7]
```diff
@@ -279,29 +279,36 @@ export class ChatApp extends React.Component<
...
+            if (created === false) {
                 this.channel = await this.chatClient.createChannel({
-                    uniqueName:
-                        this.props.username + " - " + this.props.candidateName,
+                    uniqueName: channelName,
                 });
-            } catch (error) {
+                await this.channel.add(this.props.candidateName);
+                await this.channel.add(this.props.username);
+                await Promise.all(admins.map((a) => this.channel.add(a)));
```
Comment (brikou) [inline components/ChatApp.tsx]: use this...

```js
await Promise.all([
    this.props.candidateName,
    this.props.username,
    ...admins.map((a) => this.channel.add(a)),
]);
```

File: lib/employers.ts [hunk 8]
```diff
@@ -2,5 +2,6 @@
 const employers = new Array();
```
Comment (brikou) [inline lib/employers.ts]: export const employers = [{id: "10", username: "elodie"}, /*...*/]

// never export default if you can

File: pages/_error.tsx [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import * as React from "react";
+
+class Error extends React.Component<{ statusCode: number }> {
```
Comment (brikou) [inline pages/_error.tsx]: can you use a stateless component instead

File: pages/_error.tsx [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    public static getInitialProps({ res, err }: { res: any; err: any }) {
```
Comment (brikou) [inline pages/_error.tsx]: please use correct type for res and err (see next.d.ts res should be http.IncomingMessage maybe... give a check)

File: pages/index.tsx [hunk 11]
```diff
@@ -6,14 +6,21 @@ import { Nav } from "../components/Nav";
...
 const IndexPage: React.SFC<{
-    candidateName: string;
+    candidateName: string | undefined;
     role: string;
     token: string;
     username: string;
     user_id: number;
 }> = ({ candidateName, role, token, username, user_id }) => {
+    if (!token) {
+        return <Error statusCode={404} />;
```
Comment (brikou) [inline pages/index.tsx]: can we just throw an error?

throw new Error(/ * qsdkjqsd * /)

File: pages/index.tsx [hunk 12]
```diff
@@ -29,17 +36,8 @@ const IndexPage: React.SFC<{
...
-}: {
-    query: {
-        candidate_id: number;
-        role: string;
-        user_id: number;
-    };
-}) => {
-    /* Penser à gerer erreurs 404 et 500 quand query pas bonne */
-
+(IndexPage as any).getInitialProps = async (ctx: any) => {
```
Comment (brikou) [inline pages/index.tsx]: Can you type ctx

**Review by brikou — CHANGES_REQUESTED** [review]

File: components/ChatApp.tsx [hunk 13]
```diff
@@ -278,30 +278,38 @@ export class ChatApp extends React.Component<
...
             // Si un candidate a été invité
             const previousChannel = this.channel || undefined;
-            try {
+            // false === channel non existant, true === channel déjà crée
+            let created = false;
+            const paginator = await this.chatClient.getSubscribedChannels(
+                undefined,
+            );
+            created = paginator.items.some(
+                (value: any, index: number, array: any) => {
```
Comment (brikou) [inline components/ChatApp.tsx]: rename value to channel, remove unused args, but before, config tslint to display warning on unused params

File: components/ChatApp.tsx [hunk 14]
```diff
@@ -278,30 +278,38 @@ export class ChatApp extends React.Component<
                 channels: [...prevState.channels, channel.uniqueName],
             }));
         });
-        if (this.props.candidateName !== undefined) {
+        if (this.props.candidateName !== "") {
+            const channelName =
+                this.props.username + " - " + this.props.candidateName;
```
Comment (brikou) [inline components/ChatApp.tsx]: use
```js
[this.props.username,this.props.candidateName].toString()
```

File: pages/_error.tsx [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+        err,
+    }: {
+        res: http.ServerResponse;
+        err: any;
+    }) {
+        const statusCode = res ? res.statusCode : err ? err.statusCode : null;
+        return { statusCode };
+    }
+
+    public props: any;
```
Comment (brikou) [inline pages/_error.tsx]: no already defined by react in parent component class

File: pages/_error.tsx [hunk 16]
```diff
@@ -0,0 +1,29 @@
+import * as http from "http";
+import * as React from "react";
+
+export default class Error extends React.Component {
```
Comment (brikou) [inline pages/_error.tsx]: Add props definition <{statusCode: number}>

File: pages/index.tsx [hunk 17]
```diff
@@ -1,19 +1,27 @@
...
 
 const IndexPage: React.SFC<{
-    candidateName: string;
+    candidateName: string | undefined;
     role: string;
     token: string;
     username: string;
     user_id: number;
 }> = ({ candidateName, role, token, username, user_id }) => {
+    if (!token) {
```
Comment (brikou) [inline pages/index.tsx]: I really don't like generic "!" be explicit, is it undefined or null or both, or false or ""

File: pages/index.tsx [hunk 18]
```diff
@@ -29,34 +37,46 @@ const IndexPage: React.SFC<{
...
-    } else if (role === "candidate") {
-        username = candidates[user_id];
+(IndexPage as any).getInitialProps = async (ctx: Context) => {
+    const {
+        query: { candidate_id, role, user_id },
+    }: { query: { candidate_id: string; role: string; user_id: string } } = ctx;
+    let user;
+    switch (role) {
+        case "admin":
+            user = admins.find((admin) => {
```
Comment (brikou) [inline pages/index.tsx]: use arrow func

user = admins.find((admin) => admin.id===user_id)

File: pages/index.tsx [hunk 19]
```diff
@@ -29,34 +37,46 @@ const IndexPage: React.SFC<{
...
+                return admin.id === user_id;
+            });
+            break;
+        case "employer":
+            user = employers.find((employer) => {
```
Comment (brikou) [inline pages/index.tsx]: same =>

File: pages/index.tsx [hunk 20]
```diff
@@ -29,34 +37,46 @@ const IndexPage: React.SFC<{
...
+                return employer.id === user_id;
+            });
+            break;
+        case "candidate":
+            user = candidates.find((candidat) => {
```
Comment (brikou) [inline pages/index.tsx]: same =>

File: pages/index.tsx [hunk 21]
```diff
@@ -29,34 +37,46 @@ const IndexPage: React.SFC<{
...
+                return candidat.id === user_id;
+            });
+            break;
+        default:
+            throw new Error("Wrong role in query");
```
Comment (brikou) [inline pages/index.tsx]: which role String(role)

File: pages/index.tsx [hunk 22]
```diff
@@ -29,34 +37,46 @@ const IndexPage: React.SFC<{
...
+    }
+    if (user) {
+        const candidate = candidates.find((candidat) => {
```
Comment (brikou) [inline pages/index.tsx]: =>

File: pages/index.tsx [hunk 23]
```diff
@@ -29,34 +37,46 @@ const IndexPage: React.SFC<{
...
+            return candidat.id === candidate_id;
+        });
+        const candidateName = candidate ? candidate.username : "";
```
Comment (brikou) [inline pages/index.tsx]: explicit test

File: pages/index.tsx [hunk 24]
```diff
@@ -29,34 +37,46 @@ const IndexPage: React.SFC<{
...
+        const username = user.username;
+        const { token } = await fetch(
+            `http://localhost:3000/token/${user.username}/${role}`,
+            {
+                method: "post",
+            },
+        ).then((response) => response.json());
+        return { candidateName, role, token, username, user_id };
```
Comment (brikou) [inline pages/index.tsx]: line feed before

File: server.ts [hunk 25]
```diff
@@ -38,11 +39,11 @@ app.prepare().then(() => {
     const koa = new Koa();
     koa.use(koaBodyParser());
 
-    router.get("*", async (ctx: any) => {
+    router.get("*", async (ctx: Context) => {
```
Comment (brikou) [inline server.ts]: you shouldn't need context typehint

File: pages/index.tsx [hunk 26]
```diff
@@ -29,34 +37,46 @@ const IndexPage: React.SFC<{
...
+    } else {
+        throw new Error("Wrong user id");
```
Comment (brikou) [inline pages/index.tsx]: status code 403, or 4xx

**Review by brikou — CHANGES_REQUESTED** [review]

**Comment by brikou** [comment]
ok

**Pull request merged**